File: benchmarks/fibo_tests/fibonacci.cpp
/*
Copyright (c) 2019 Oracle

The Universal Permissive License (UPL), Version 1.0

Subject to the condition set forth below, permission is hereby granted to any person
obtaining a copy of this software, associated documentation and/or data (collectively the
"Software"), free of charge and under any and all copyright rights in the Software, and
any and all patent rights owned or freely licensable by each licensor hereunder covering
either (i) the unmodified Software as contributed to or provided by such licensor, or (ii)
the Larger Works (as defined below), to deal in both

(a) the Software, and

(b) any piece of software and/or hardware listed in the lrgrwrks.txt file if one is
included with the Software (each a “Larger Work” to which the Software is contributed by
such licensors),

without restriction, including without limitation the rights to copy, create derivative
works of, display, perform, and distribute the Software and make, use, sell, offer for
sale, import, export, have made, and have sold the Software and the Larger Work(s), and to
sublicense the foregoing rights on either these or other terms.

This license is subject to the following condition:

The above copyright notice and either this complete permission notice or at a minimum a
reference to the UPL must be included in all copies or substantial portions of the
Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
DEALINGS IN THE SOFTWARE.
*/

#include "Blackhole.hpp"

//@BenchmarkMode(Mode.AverageTime)
//@OutputTimeUnit(TimeUnit.NANOSECONDS)
//@Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
//@Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
//@Fork(1)
@State(Scope.Benchmark)
class FibonacciTests {
public:

   //@State(Scope.Thread)
   //class InputValue {
   //public:
   //   //@Param({"1", "2", "3"})
   //   @Param({"1"})
   //   static int rep;
   //
   //   @Param({"25"})
   //   volatile static long number;
   //   //static long number;
   //};

  @State(Scope.Benchmark)
  struct Values {
     int a{0};
     int b{0};
     int c{0};
     int d{0};
     int e{0};
     //int f{0};
  };

  @Setup(Level.Iteration)
  void setup(Values &vals) {
     vals.a = 0x10023a41;
     vals.b = 0x0139fc27;
  }

private:

public:
   @BenchmarkMode(Mode.AverageTime)
   @OutputTimeUnit(TimeUnit.NANOSECONDS)
   @Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
   @Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
   @Fork(1)
   @Benchmark
   static int fibonacci_local(Values &vals) {
      int a = vals.a;
      int b = vals.b;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      return b;
   }

   @BenchmarkMode(Mode.AverageTime)
   @OutputTimeUnit(TimeUnit.NANOSECONDS)
   @Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
   @Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
   @Fork(1)
   @Benchmark
   static int fibonacci_nonlocal(Values &vals) {
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      return vals.b;
   }

   @BenchmarkMode(Mode.AverageTime)
   @OutputTimeUnit(TimeUnit.NANOSECONDS)
   @Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
   @Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
   @Fork(1)
   @Benchmark
   static int fibonacci_para2(Values &vals) {
      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      return vals.b;
   }

   @BenchmarkMode(Mode.AverageTime)
   @OutputTimeUnit(TimeUnit.NANOSECONDS)
   @Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
   @Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
   @Fork(1)
   @Benchmark
   static int fibonacci_para3(Values &vals) {
      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.e = (vals.a << 1) + 3*vals.b;

      vals.a = vals.d + vals.e;
      vals.b = vals.d + (vals.e << 1);
      vals.c = (vals.d << 1) + 3*vals.e;

      vals.d = vals.b + vals.c;
      vals.e = vals.b + (vals.c << 1);
      vals.a = (vals.b << 1) + 3*vals.c;

      vals.b = vals.e + vals.a;
      vals.c = vals.e + (vals.a << 1);
      vals.d = (vals.e << 1) + 3*vals.a;

      vals.e = vals.c + vals.d;
      vals.a = vals.c + (vals.d << 1);
      vals.b = (vals.c << 1) + 3*vals.d;

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.e = (vals.a << 1) + 3*vals.b;

      vals.a = vals.d + vals.e;
      vals.b = vals.d + (vals.e << 1);
      vals.c = (vals.d << 1) + 3*vals.e;

      vals.d = vals.b + vals.c;
      vals.e = vals.b + (vals.c << 1);
      vals.a = (vals.b << 1) + 3*vals.c;

      vals.b = vals.e + vals.a;
      vals.c = vals.e + (vals.a << 1);
      vals.d = (vals.e << 1) + 3*vals.a;

      vals.e = vals.c + vals.d;
      vals.a = vals.c + (vals.d << 1);
      vals.b = (vals.c << 1) + 3*vals.d;

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.e = (vals.a << 1) + 3*vals.b;

      vals.a = vals.d + vals.e;
      vals.b = vals.d + (vals.e << 1);
      vals.c = (vals.d << 1) + 3*vals.e;

      vals.d = vals.b + vals.c;
      vals.e = vals.b + (vals.c << 1);
      vals.a = (vals.b << 1) + 3*vals.c;

      vals.b = vals.e + vals.a;
      vals.c = vals.e + (vals.a << 1);
      vals.d = (vals.e << 1) + 3*vals.a;

      vals.e = vals.c + vals.d;
      vals.a = vals.c + (vals.d << 1);
      vals.b = (vals.c << 1) + 3*vals.d;

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.e = (vals.a << 1) + 3*vals.b;

      vals.a = vals.d + vals.e;
      vals.b = vals.d + (vals.e << 1);
      vals.c = (vals.d << 1) + 3*vals.e;

      vals.d = vals.b + vals.c;
      vals.e = vals.b + (vals.c << 1);
      vals.a = (vals.b << 1) + 3*vals.c;

      vals.b = vals.e + vals.a;
      vals.c = vals.e + (vals.a << 1);
      vals.d = (vals.e << 1) + 3*vals.a;

      vals.e = vals.c + vals.d;
      vals.a = vals.c + (vals.d << 1);
      vals.b = (vals.c << 1) + 3*vals.d;

      return vals.b;
   }

};

//int FibonacciTests::InputValue::rep;
//volatile long FibonacciTests::InputValue::number = 25;
//long FibonacciTests::InputValue::number = 25;

File: benchmarks/fibo_tests/Fibonacci.java
/*
 * Copyright (c) 2019, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

//package org.graalvm.benchmarks;
package leick;

import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Fork;
import org.openjdk.jmh.annotations.Level;
import org.openjdk.jmh.annotations.Measurement;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.annotations.Param;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.Setup;
import org.openjdk.jmh.annotations.State;
//import org.graalvm.polyglot.*;
import org.openjdk.jmh.annotations.TearDown;
import org.openjdk.jmh.annotations.Warmup;
import org.openjdk.jmh.infra.Blackhole;

//import java.io.File;
//import java.io.IOException;
import java.util.concurrent.TimeUnit;


@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
@Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
@Fork(1)
public class Fibonacci {

    //@State(Scope.Thread)
    //public static class InputValue {
    //    volatile static long number = 25;
    //}

    @State(Scope.Benchmark)
    public static class Values {
	int a = 0;
	int b = 0;
	int c = 0;
	int d = 0;
	int e = 0;
	//int f = 0;

	@Setup(Level.Iteration)
        public void setup() {
	    a = 0x10023a41;
	    b = 0x0139fc27;
	}
    };


   @Benchmark
   public int fibonacci_local() {
      int a = 0x10023a41;
      int b = 0x0139fc27;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      a += b;
      b += a;
      return b;
   }

   @Benchmark
   public int fibonacci_nonlocal(Values vals) {
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      vals.a += vals.b;
      vals.b += vals.a;
      return vals.b;
   }

   @Benchmark
   public int fibonacci_para2(Values vals) {
      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.a = vals.c + vals.d;
      vals.b = vals.c + (vals.d << 1);

      return vals.b;
   }

   @Benchmark
   public int fibonacci_para3(Values vals) {
      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.e = (vals.a << 1) + 3*vals.b;

      vals.a = vals.d + vals.e;
      vals.b = vals.d + (vals.e << 1);
      vals.c = (vals.d << 1) + 3*vals.e;

      vals.d = vals.b + vals.c;
      vals.e = vals.b + (vals.c << 1);
      vals.a = (vals.b << 1) + 3*vals.c;

      vals.b = vals.e + vals.a;
      vals.c = vals.e + (vals.a << 1);
      vals.d = (vals.e << 1) + 3*vals.a;

      vals.e = vals.c + vals.d;
      vals.a = vals.c + (vals.d << 1);
      vals.b = (vals.c << 1) + 3*vals.d;

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.e = (vals.a << 1) + 3*vals.b;

      vals.a = vals.d + vals.e;
      vals.b = vals.d + (vals.e << 1);
      vals.c = (vals.d << 1) + 3*vals.e;

      vals.d = vals.b + vals.c;
      vals.e = vals.b + (vals.c << 1);
      vals.a = (vals.b << 1) + 3*vals.c;

      vals.b = vals.e + vals.a;
      vals.c = vals.e + (vals.a << 1);
      vals.d = (vals.e << 1) + 3*vals.a;

      vals.e = vals.c + vals.d;
      vals.a = vals.c + (vals.d << 1);
      vals.b = (vals.c << 1) + 3*vals.d;

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.e = (vals.a << 1) + 3*vals.b;

      vals.a = vals.d + vals.e;
      vals.b = vals.d + (vals.e << 1);
      vals.c = (vals.d << 1) + 3*vals.e;

      vals.d = vals.b + vals.c;
      vals.e = vals.b + (vals.c << 1);
      vals.a = (vals.b << 1) + 3*vals.c;

      vals.b = vals.e + vals.a;
      vals.c = vals.e + (vals.a << 1);
      vals.d = (vals.e << 1) + 3*vals.a;

      vals.e = vals.c + vals.d;
      vals.a = vals.c + (vals.d << 1);
      vals.b = (vals.c << 1) + 3*vals.d;

      vals.c = vals.a + vals.b;
      vals.d = vals.a + (vals.b << 1);
      vals.e = (vals.a << 1) + 3*vals.b;

      vals.a = vals.d + vals.e;
      vals.b = vals.d + (vals.e << 1);
      vals.c = (vals.d << 1) + 3*vals.e;

      vals.d = vals.b + vals.c;
      vals.e = vals.b + (vals.c << 1);
      vals.a = (vals.b << 1) + 3*vals.c;

      vals.b = vals.e + vals.a;
      vals.c = vals.e + (vals.a << 1);
      vals.d = (vals.e << 1) + 3*vals.a;

      vals.e = vals.c + vals.d;
      vals.a = vals.c + (vals.d << 1);
      vals.b = (vals.c << 1) + 3*vals.d;

      return vals.b;
   }
}

